cocos/core/pipeline/multiLightDSManager.ts

```diff
@@ -0,0 +1,110 @@
+ * @module pipeline
+ */
+
+import { RenderPipeline } from './render-pipeline';
+import { Device, BufferUsageBit, MemoryUsageBit, BufferInfo,
+    Filter, Address, Sampler, DescriptorSet, DescriptorSetInfo } from '../gfx';
+import { Light } from '../renderer/scene/light';
+import { UBOGlobal, UBOShadow, UBOCamera, UNIFORM_SHADOWMAP_BINDING,
+    UNIFORM_SPOT_LIGHTING_MAP_TEXTURE_BINDING, PipelineGlobalBindings } from './define';
+import { genSamplerHash, samplerLib } from '../renderer/core/sampler-lib';
+import { builtinResMgr } from '../builtin/builtin-res-mgr';
+import { Texture2D } from '../assets/texture-2d';
+
+const _samplerInfo = [
+    Filter.LINEAR,
+    Filter.LINEAR,
+    Filter.NONE,
+    Address.CLAMP,
+    Address.CLAMP,
+    Address.CLAMP,
+];
+
+export class MultiLightDSManager {
+    private _pipeline: RenderPipeline;
+    private _device: Device;
+    private _descriptorSetMap: Map<Light, DescriptorSet> = new Map();
+    private _sampler: Sampler;
+
+    constructor (pipeline: RenderPipeline) {
+        this._pipeline = pipeline;
+        this._device = pipeline.device;
+
+        const shadowMapSamplerHash = genSamplerHash(_samplerInfo);
+        this._sampler = samplerLib.getSampler(this._device, shadowMapSamplerHash);
+    }
+
+    public updateDescriptorSet (light: Light) {
+        const descriptorSet = this._descriptorSetMap.get(light);
+        if (descriptorSet) {
+            const globalDescriptorSet = this._pipeline.descriptorSet;
+            for (let i = PipelineGlobalBindings.SAMPLER_SHADOWMAP; i < PipelineGlobalBindings.COUNT; i++) {
+                if (globalDescriptorSet.getTexture(i)) {
+                    descriptorSet.bindSampler(i, globalDescriptorSet.getSampler(i));
+                    descriptorSet.bindTexture(i, globalDescriptorSet.getTexture(i));
+                }
+            }
+
+            descriptorSet.update();
+        }
+    }
+
+    public getOrCreateDescriptorSet (light: Light) {
+        if (!this._descriptorSetMap.has(light)) {
+            const device = this._device;
+            const descriptorSet = device.createDescriptorSet(new DescriptorSetInfo(this._pipeline.descriptorSetLayout));
+
+            descriptorSet.bindBuffer(UBOGlobal.BINDING, this._pipeline.descriptorSet.getBuffer(UBOGlobal.BINDING));
+
+            descriptorSet.bindBuffer(UBOCamera.BINDING, this._pipeline.descriptorSet.getBuffer(UBOCamera.BINDING));
+
+            const shadowBUO = device.createBuffer(new BufferInfo(
+                BufferUsageBit.UNIFORM | BufferUsageBit.TRANSFER_DST,
+                MemoryUsageBit.HOST | MemoryUsageBit.DEVICE,
+                UBOShadow.SIZE,
+                UBOShadow.SIZE,
+            ));
+            descriptorSet.bindBuffer(UBOShadow.BINDING, shadowBUO);
+
+            descriptorSet.bindSampler(UNIFORM_SHADOWMAP_BINDING, this._sampler);
+            descriptorSet.bindTexture(UNIFORM_SHADOWMAP_BINDING, builtinResMgr.get<Texture2D>('default-texture').getGFXTexture()!);
```

Comment:
Use defaultTexture._uuid

Reply:
The engine runs like this using

Reply:
My concern is that if the uuid changes this code would break. So directly referencing the uuid would be better.